src/protections/ProtectionManager.ts

```diff
@@ -0,0 +1,393 @@
+import { Protection } from "./IProtection";
+import { BasicFlooding } from "./BasicFlooding";
+import { DetectFederationLag } from "./DetectFederationLag";
+import { WordList } from "./WordList";
+import { MessageIsVoice } from "./MessageIsVoice";
+import { MessageIsMedia } from "./MessageIsMedia";
+import { TrustedReporters } from "./TrustedReporters";
+import { JoinWaveShortCircuit } from "./JoinWaveShortCircuit";
+import { Mjolnir } from "../Mjolnir";
+import { extractRequestError, LogLevel, LogService, Permalinks } from "matrix-bot-sdk";
+import { ProtectionSettingValidationError } from "./ProtectionSettings";
+import { Consequence, ConsequenceType } from "./consequence";
+import { htmlEscape } from "../utils";
+import { ERROR_KIND_FATAL, ERROR_KIND_PERMISSION } from "../ErrorCache";
+import { RoomUpdateError } from "../models/RoomUpdateError";
+
+const PROTECTIONS: Protection[] = [
+    new FirstMessageIsImage(),
+    new BasicFlooding(),
+    new WordList(),
+    new MessageIsVoice(),
+    new MessageIsMedia(),
+    new TrustedReporters(),
+    new DetectFederationLag(),
+    new JoinWaveShortCircuit(),
+];
+
+const ENABLED_PROTECTIONS_EVENT_TYPE = "org.matrix.mjolnir.enabled_protections";
+const CONSEQUENCE_EVENT_DATA = "org.matrix.mjolnir.consequence";
+
+export class ProtectionManager {
+    private _protections = new Map<string /* protection name */, Protection>();
+    get protections(): Readonly<Map<string /* protection name */, Protection>> {
+        return this._protections;
+    }
+
+
+    constructor(private readonly mjolnir: Mjolnir) {
+    }
+
+    /*
+     * Take all the builtin protections, register them to set their enabled (or not) state and
+     * update their settings with any saved non-default values
+     */
+    public async start() {
+        this.mjolnir.reportManager.on("report.new", this.handleReport.bind(this));
+        this.mjolnir.client.on("room.event", this.handleEvent.bind(this));
+        for (const protection of PROTECTIONS) {
+            try {
+                await this.registerProtection(protection);
+            } catch (e) {
+                this.mjolnir.managementRoom.logMessage(LogLevel.WARN, "ProtectionManager", extractRequestError(e));
+            }
+        }
+    }
+
+    /*
+     * Given a protection object; add it to our list of protections, set whether it is enabled
+     * and update its settings with any saved non-default values.
+     *
+     * @param protection The protection object we want to register
+     */
+    public async registerProtection(protection: Protection) {
+        this._protections.set(protection.name, protection)
+
+        let enabledProtections: { enabled: string[] } | null = null;
+        try {
+            enabledProtections = await this.mjolnir.client.getAccountData(ENABLED_PROTECTIONS_EVENT_TYPE);
+        } catch {
+            // this setting either doesn't exist, or we failed to read it (bad network?)
+            // TODO: retry on certain failures?
```

Comment:
Not sure what you mean by this?

Reply:
Nothing specific right now.

However, now that we have concluded that we're staying with matrix-bot-sdk, I believe that we should write a layer on top of (some selected) state events and `getAccountData` to be event-based and implement caching. This will mean that our future replacement for `getAccountData` will be sync and will never fail.

Reply:
Filed as https://github.com/matrix-org/mjolnir/issues/374 .